Multi-threaded quantization

ikawrakow opened the pull request:
This PR adds multi-threading for quantization.

The gain is very minor for small models (e.g., LLaMA 7B) and simple quantization (`Q4_0` and `Q4_1`), but very significant for large models and the now more elaborate `Q4_2` quantization.

`quantize-stats` now finishes in just 14.5 seconds (7B) or 44 seconds (13B) on my computer for all 3 quantization types. The single-threaded version took 144 seconds (7B) or 242 seconds (13B).

Commit:
Multi-threading quantization.

Not much gain for simple quantizations, bit it will be important
for quantizations that require more CPU cycles.

Commit:
Multi-threading for quantize-stats

It now does the job in ~14 seconds on my Mac for
Q4_0, Q4_1 and Q4_2. Single-threaded it was taking
more than 2 minutes after adding the more elaborate
version of Q4_2.

DannyDaemonic commented:
This could make more accurate but slow quantization methods more practical. (See #835.)

ggerganov commented on llama.cpp:
```diff
@@ -1578,6 +1581,8 @@ static void llama_model_quantize_internal(const std::string & fname_inp, const s
         default: throw format("invalid output file type %d\n", ftype);
     };
 
+    if (nthread <= 0) nthread = std::thread::hardware_concurrency();
+
```
```suggestion
    if (nthread <= 0) {
        nthread = std::thread::hardware_concurrency();
    }
    
    fprintf(stderr, "%s : using %d threads\n", nthread);
```

ggerganov commented on llama.cpp:
```diff
@@ -1639,21 +1647,37 @@ static void llama_model_quantize_internal(const std::string & fname_inp, const s
-                    } break;
-                default:
-                    LLAMA_ASSERT(false);
+            int chunk_size = 32 * 512;
+            int nchunk = (nelements + chunk_size - 1)/chunk_size;
+            int nthread_use = nthread > 1 ? std::max(1, std::min(nthread, nchunk)) : 1;
```
```suggestion
            const int chunk_size = 32 * 512;
            const int nchunk = (nelements + chunk_size - 1)/chunk_size;
            const int nthread_use = nthread > 1 ? std::max(1, std::min(nthread, nchunk)) : 1;
```

ggerganov commented on llama.cpp:
```diff
@@ -1639,21 +1647,37 @@ static void llama_model_quantize_internal(const std::string & fname_inp, const s
+                        local_size += ggml_quantize_chunk(new_type, f32_data, new_data, first, last - first, local_hist.data());
+                    }
+                };
+                if (int(workers.size()) < nthread_use-1) workers.resize(nthread_use-1);
+                for (int it=0; it<nthread_use-1; ++it) workers[it] = std::thread(compute);
+                compute();
+                for (int it=0; it<nthread_use-1; ++it) workers[it].join();
```
```suggestion
                if (int(workers.size()) < nthread_use - 1) workers.resize(nthread_use - 1);
                for (int it = 0; it < nthread_use - 1; ++it) workers[it] = std::thread(compute);
                compute();
                for (int it = 0; it < nthread_use - 1; ++it) workers[it].join();
```

ggerganov commented on ggml.c:
```diff
@@ -11870,6 +11870,33 @@ size_t ggml_quantize_q4_2(const float * src, void * dst, int n, int k, int64_t *
+    switch (type) {
+        case GGML_TYPE_Q4_0:
+            {
+                assert (start % QK4_0 == 0);
+                block_q4_0 * block = (block_q4_0*)dst + start / QK4_0;
+                result = ggml_quantize_q4_0(src + start, block, n, n, hist);
+            } break;
+        case GGML_TYPE_Q4_1:
+            {
+                assert (start % QK4_1 == 0);
+                block_q4_1 * block = (block_q4_1*)dst + start / QK4_1;
+                result = ggml_quantize_q4_1(src + start, block, n, n, hist);
+            } break;
+        case GGML_TYPE_Q4_2:
+            {
+                assert (start % QK4_2 == 0);
+                block_q4_2 * block = (block_q4_2*)dst + start / QK4_2;
+                result = ggml_quantize_q4_2(src + start, block, n, n, hist);
+            } break;
+        default:
+            assert(false);
```
Change these `assert` to `GGML_ASSERT` so it crashes even in Release if the user provides wrong data

Commit:
Reviewer comments

Commit:
Avoiding compiler confusion

After changing chunk_size to const int as suggested by
@ggerganov, clang and GCC starting to warn me that I don't
need to capture it in the lambda. So, I removed it from the
capture list. But that makes the MSVC build fail. So,
making it a constexpr to make every compiler happy.

Commit:
Still fighting with lambda captures in MSVC

prusnak commented:
Please resolve conflicts with the master branch

ggerganov committed:
Merge branch 'master' into multi-thread-quantize
